### PR TITLE
feat: LT-2118 agree to declaration when submitting

### PR DESCRIPTION
**Pull request by rorysproule1**

### CODE REVIEW

File: applications/models.py [hunk 1]
```diff
@@ -157,6 +157,7 @@ class BaseApplication(ApplicationPartyMixin, Case):
     compliant_limitations_eu_ref = models.CharField(default=None, blank=True, null=True, max_length=2200)
 
     intended_end_use = models.CharField(default=None, blank=True, null=True, max_length=2200)
+    is_agreed_to_foi = models.BooleanField(blank=True, default=None, null=True)
```
Comment (jfaracik) [inline applications/models.py]: Could this just be "agreed_to_foi"? "agreed_to_foi" already sounds like it returns a boolean so the "is_" is superfluous

Comment (rorysproule1) [reply applications/models.py]: Updated to agreed_to_foi. Thanks :)

File: applications/libraries/case_status_helpers.py [hunk 2]
```diff
@@ -7,3 +11,10 @@ def get_case_statuses(read_only):
         return CaseStatusEnum.read_only_statuses()
     else:
         return [status for status, value in CaseStatusEnum.choices if not CaseStatusEnum.is_read_only(status)]
+
+
+def set_application_sla(application):
+    application.status = get_case_status_by_status(CaseStatusEnum.SUBMITTED)
+    application.submitted_at = timezone.now()
+    application.sla_remaining_days = get_application_target_sla(application.case_type.sub_type)
+    application.sla_days = 0
```
Comment (BenAndrew-Kainos) [inline applications/libraries/case_status_helpers.py]: should have a .save() here

File: applications/tests/tests_mod_clearance_submit.py [hunk 3]
```diff
@@ -81,15 +84,46 @@ def test_submit_exhibition_clearance_without_details_failure(self):
...
+        }
+
+        url = reverse("applications:application_submit", kwargs={"pk": self.draft.id})
+        response = self.client.put(url, data, **self.exporter_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        application = ExhibitionClearanceApplication.objects.get()
+        self.assertIsNotNone(application.submitted_at)
+        self.assertEqual(application.status.status, CaseStatusEnum.SUBMITTED)
```
Comment (BenAndrew-Kainos) [inline applications/tests/tests_mod_clearance_submit.py]: check agreed_to_foi

File: applications/tests/tests_mod_clearance_submit.py [hunk 4]
```diff
@@ -150,7 +188,38 @@ def test_submit_gifting_clearance_with_location_failure(self):
...
+        }
+
+        url = reverse("applications:application_submit", kwargs={"pk": self.draft.id})
+        response = self.client.put(url, data, **self.exporter_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        application = GiftingClearanceApplication.objects.get()
+        self.assertIsNotNone(application.submitted_at)
+        self.assertEqual(application.status.status, CaseStatusEnum.SUBMITTED)
```
Comment (BenAndrew-Kainos) [inline applications/tests/tests_mod_clearance_submit.py]: check agreed_to_foi

File: applications/tests/tests_mod_clearance_submit.py [hunk 5]
```diff
@@ -246,4 +321,37 @@ def test_submit_F680_clearance_without_end_use_details_failure(self):
...
+        }
+
+        url = reverse("applications:application_submit", kwargs={"pk": self.draft.id})
+        response = self.client.put(url, data, **self.exporter_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        application = F680ClearanceApplication.objects.get()
+        self.assertIsNotNone(application.submitted_at)
+        self.assertEqual(application.status.status, CaseStatusEnum.SUBMITTED)
```
Comment (BenAndrew-Kainos) [inline applications/tests/tests_mod_clearance_submit.py]: check agreed_to_foi

File: applications/tests/tests_open_application_submit.py [hunk 6]
```diff
@@ -61,3 +61,35 @@ def test_submit_open_application_without_end_use_details_failure(self):
...
+
+        url = reverse("applications:application_submit", kwargs={"pk": self.draft.id})
+        response = self.client.put(url, data, **self.exporter_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        case = Case.objects.get()
+        self.assertEqual(case.id, self.draft.id)
+        self.assertIsNotNone(case.submitted_at)
+        self.assertEqual(case.status.status, CaseStatusEnum.SUBMITTED)
```
Comment (BenAndrew-Kainos) [inline applications/tests/tests_open_application_submit.py]: check agreed_to_foi

File: applications/tests/tests_standard_application_submit.py [hunk 7]
```diff
@@ -181,8 +181,11 @@ def test_exp_set_application_status_to_submitted_when_previously_applicant_editi
...
 
+        data = {"submit_declaration": True, "agreed_to_declaration": True, "agreed_to_foi": True}
+
         url = reverse("applications:application_submit", kwargs={"pk": standard_application.id})
-        response = self.client.put(url, **self.exporter_headers)
+
+        response = self.client.put(url, data, **self.exporter_headers)
 
         standard_application.refresh_from_db()
         self.assertEqual(response.status_code, status.HTTP_200_OK)
```
Comment (BenAndrew-Kainos) [inline applications/tests/tests_standard_application_submit.py]: check agreed_to_foi

File: applications/tests/tests_standard_application_submit.py [hunk 8]
```diff
@@ -235,13 +238,42 @@ def test_submit_standard_application_without_end_use_details_failure(self):
...
+        data = {"submit_declaration": True, "agreed_to_declaration": True, "agreed_to_foi": True}
+
+        url = reverse("applications:application_submit", kwargs={"pk": self.draft.id})
+        response = self.client.put(url, data, **self.exporter_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        case = Case.objects.get(id=self.draft.id)
+        self.assertIsNotNone(case.submitted_at)
+        self.assertEqual(case.status.status, CaseStatusEnum.SUBMITTED)
```
Comment (BenAndrew-Kainos) [inline applications/tests/tests_standard_application_submit.py]: check agreed_to_foi

File: applications/views/applications.py [hunk 9]
```diff
@@ -274,43 +276,60 @@ class ApplicationSubmission(APIView):
...
             )
-        previous_application_status = application.status
+
+        if application.case_type.sub_type in [
+            CaseTypeSubTypeEnum.HMRC,
+            CaseTypeSubTypeEnum.EUA,
+            CaseTypeSubTypeEnum.GOODS,
+        ]:
+            set_application_sla(application)
+            application.save()
```
Comment (BenAndrew-Kainos) [inline applications/views/applications.py]: Please move .save() into set_application_sla

File: applications/views/applications.py [hunk 10]
```diff
@@ -274,43 +276,60 @@ class ApplicationSubmission(APIView):
...
+        ]:
+            if request.data.get("submit_declaration"):
+                errors = _validate_agree_to_declaration(request, errors)
+                if errors:
+                    return JsonResponse(data={"errors": errors}, status=status.HTTP_400_BAD_REQUEST)
+
+                # If a valid declaration is provided, save the application
+                application.agreed_to_foi = request.data.get("agreed_to_foi")
+                set_application_sla(application)
+                application.save()
```
Comment (BenAndrew-Kainos) [inline applications/views/applications.py]: Please move .save() into set_application_sla

File: applications/libraries/case_status_helpers.py (hunk 2 above)
Comment (rorysproule1) [reply applications/libraries/case_status_helpers.py]: Added application.save to set_application_sla. Cheers!

File: applications/creators.py [hunk 11]
```diff
@@ -156,17 +156,29 @@ def _validate_end_use_details(draft, errors, application_type):
...
-            errors["end_use_details"] = strings.Applications.Generic.NO_END_USE_DETAILS
+            errors["end_use_details"] = [strings.Applications.Generic.NO_END_USE_DETAILS]
+
+    return errors
+
+
+def _validate_agree_to_declaration(request, errors):
+    """ Checks the exporter has agreed to the T&Cs of the licence """
+
+    if str_to_bool(request.data.get("agreed_to_declaration")) is not True:
```
Comment (a-bek) [inline applications/creators.py]: `if not str_to_bool(...):`

File: applications/libraries/case_status_helpers.py (hunk 2 above)
Comment (andytaylorkainos) [reply applications/libraries/case_status_helpers.py]: can you bunch the FOI details in here too please if youre saving within this method :)

File: applications/creators.py (hunk 11 above)
Comment (rorysproule1) [reply applications/creators.py]: Amended, cheers!

File: applications/tests/tests_open_application_submit.py (hunk 6 above)
Comment (rorysproule1) [reply applications/tests/tests_open_application_submit.py]: Added check to each of your queries

**Pull request merged**